lib/client/client.js

```diff
@@ -1014,6 +1014,8 @@ Client.prototype.connect = function connect () {
       self.emit('connectTimeout', err)
     } else if (err.code === 'ECONNREFUSED') {
       self.emit('connectRefused', err)
+    } else if (err.code === 'CERT_HAS_EXPIRED' || err.code === 'UNABLE_TO_VERIFY_LEAF_SIGNATURE') {
+      self.emit('connectError', err)
```

Comment:
@UziTech we already have this event at https://github.com/ldapjs/node-ldapjs/blob/8384a4736ff03bc6a09da5ee0e12d69eae02d277/lib/client/client.js#L793-L805

So this is not a brand new event.

Reply:
Yes but I don't understand why emitting those events is better than emitting an `error` event. `error` events are at least handled as errors if they are not handled by the user.

Reply:
I don't have an opinion on this. If you decide that connectError is appropriate, I will try to add some tests, otherwise I guess we can just close this PR. Either way is fine with me.

Reply:
@jonekdahl there's no reason to close the PR. Discussion is how code gets improved and accepted. 

@UziTech is referencing https://nodejs.org/dist/latest-v12.x/docs/api/events.html#events_error_events

It makes sense to fallback on that behavior because this library has a plethora of undocumented events. As a result, users are not listening for all of them correctly. Some try, but it's very difficult to do given the lack of documentation.

However, we can put a blocker on v2 that requires the documentation to be updated. I have added this PR to the project around improving documentation so that this event should not be forgotten in that work.

If we emit everything under the `error` event then the user has to do a lot of work to figure out what kind of error they have received and then determine how to handle it appropriately. At the moment, narrowing down errors by event type is the most user friendly way of doing this, I think. If we had an error system like Fastify(https://github.com/fastify/fastify/blob/7fa4bddca07aac18e0600cff40d97b95a96ab970/lib/errors.js) then I would be more inclined to agree that all errors should just emit via `error`.

What do you think @UziTech?

Reply:
I'm not saying emitting other error events is the wrong way to do it, but I'm trying to understand why it would be better. The undocumented error events seems to be a big issue for users.

I like the idea of creating our own errors like fastly and emitting them as an `error` event. If they are documented they should be just as easy to catch and they still crash the process if they aren't dealt with making them easier to debug. If someone doesn't subscribe to the `connectError` event it just fails silently right now.

Reply:
I feel like just documenting the error codes (e.g. `'ECONNREFUSED'`, `'CERT_HAS_EXPIRED'`, etc.) might be good enough. I don't think we need to go as far as creating our own errors.

Reply:
For version 2.0, perhaps the safest way to go is to keep whatever the 1.x behavior was, and document it. I would prefer to be on a stable (released) version of ldapjs and a supported Node LTS sooner rather than later, which is why I am trying to help out here.

If you agree, I can try to create a test to verify the 1.0.2 behavior on Node 8 and another test for the same behavior on the `next` branch on Node 10+.

Reply:
I agree with your assessment for the safest approach.

Please leave the `master` branch alone. Node 8 is EOL.

A test for this added to this PR would be fantastic. I think my initial idea around testing for it is the way to go. https://www.npmjs.com/package/node-forge could be helpful in this endeavor.

I truly appreciate you stepping in to help and get this library up to a new "stable" release.

Reply:
Do either of you happen to know what the 1.0.2 behavior is? If this is not a regression, fixing it should perhaps not be a blocker for a 2.0 release, right?

Reply:
The 1.0.2 behavior was to just emit an `error`. v2.0 release has many other breaking changes so I don't think it will be a problem to add one more.

Reply:
Regardless of whether or not v2 includes breaking changes, I'm am taking the stance of leaving v1 in its current known state and making v2 a sort of start fresh for the project.

Reply:
@jonekdahl can you make the change to check for listeners count on `connectError` and fallback to `error` if the count is `0`?

Reply:
@jsumners Sure, I could do that. Do you mean for connectError only, or for other custom error events like `connectTimeout`, `connectRefused` and `setupError` as well?

Reply:
Let's limit it to the error this PR resolves for now. A separate PR can cover all cases.